contracts/utils/RetireToucanCarbon.sol

```diff
@@ -0,0 +1,1405 @@
+            IERC20(wsKLIMA).safeTransfer(msg.sender, wrappedDust);
+        }
+
+        _retireCarbon(_amount, _beneficiary, _listTCO2, _toucanPool);
+
+        _bondFees(_toucanPool);
+    }
+
+    function retireWithUSDC(
+        uint256 _amount,
+        string calldata _beneficiary,
+        address[] calldata _listTCO2,
+        address _toucanPool
+    ) public {
+        require(isToucanPool[_toucanPool], "Not a Toucan Protocol Carbon Pool");
+
+        uint256 totalPoolNeeded = (_amount * (1000 + feeAmount)) / 1000;
+
+        (uint256 amountIn, address[] memory path) = getNeededUSDC(
+            _toucanPool,
+            poolRouter[_toucanPool],
+            totalPoolNeeded
+        );
+
+        // Transfer and swap from USDC to Carbon Pool
+
+        IERC20(USDC).safeTransferFrom(msg.sender, address(this), amountIn);
+
+        IERC20(USDC).safeIncreaseAllowance(poolRouter[_toucanPool], amountIn);
+
+        uint256[] memory amounts = IUniswapV2Router02(poolRouter[_toucanPool])
+            .swapTokensForExactTokens(
+                totalPoolNeeded,
+                amountIn,
+                path,
+                address(this),
+                block.timestamp
+            );
+        // Return any dust remaining (slippage protection)
+
+        uint256 tradeDust = amountIn -
+            (amounts[0] == 0 ? amounts[1] : amounts[0]);
+
+        IERC20(USDC).safeTransfer(msg.sender, tradeDust);
+
+        _retireCarbon(_amount, _beneficiary, _listTCO2, _toucanPool);
+
+        _bondFees(_toucanPool);
+    }
+
+    function retireWithPool(
+        uint256 _amount,
+        string calldata _beneficiary,
+        address[] calldata _listTCO2,
+        address _toucanPool
+    ) public {
+        require(isToucanPool[_toucanPool], "Not a Toucan Protocol Carbon Pool");
+
+        uint256 totalPoolNeeded = (_amount * (1000 + feeAmount)) / 1000;
+
+        // Transfer in the pool token to retire
+        IERC20(_toucanPool).safeTransferFrom(
+            msg.sender,
+            address(this),
+            totalPoolNeeded
+        );
+
+        // Retire the carbon
+        _retireCarbon(_amount, _beneficiary, _listTCO2, _toucanPool);
+
+        _bondFees(_toucanPool);
+    }
+
+    function _retireCarbon(
+        uint256 _amount,
+        string calldata _beneficiary,
+        address[] calldata _listTCO2,
+        address _toucanPool
+    ) internal {
+        // The carbon to be retired should be transferred or swapped inside this contract before calling this function.
+        uint256 leftToBurn = _amount;
+
+        for (uint256 i = 0; i < _listTCO2.length && leftToBurn > 0; i++) {
+            // Start with the first address in the list and work your way down.
+
+            // Get the pools balance of TCO2
+            uint256 poolBalance = IERC20(_listTCO2[i]).balanceOf(_toucanPool);
+
+            // Error check for possible 0 balance / stale lists
+            if (poolBalance == 0) {} else {
```

Comment:
Are you planning to throw an error for a zero pool balance? If not:
```suggestion
            if (poolBalance != 0) {
```

Reply:
Implemented in latest commit.